test/gtest/type-builder.cpp

```diff
@@ -1251,165 +1261,141 @@ TEST_F(TypeTest, TestTypeRelations) {
-
-  assertLUB(exactDef, exactDef, exactDef, exactDef);
-  assertLUB(exactDef, nullExactDef, nullExactDef, exactDef);
-  assertLUB(exactDef, none, def, exactNone);
-  assertLUB(exactDef, nullNone, nullDef, exactNone);
-  assertLUB(exactDef, exactNone, exactDef, exactNone);
-  assertLUB(exactDef, nullExactNone, nullExactDef, exactNone);
-  assertLUB(exactDef, func, Type(Type::none), unreachable);
-  assertLUB(exactDef, nullFunc, Type(Type::none), unreachable);
-  assertLUB(exactDef, exactFunc, Type(Type::none), unreachable);
-  assertLUB(exactDef, nullExactFunc, Type(Type::none), unreachable);
-  assertLUB(exactDef, i32, Type(Type::none), unreachable);
-  assertLUB(exactDef, unreachable, exactDef, unreachable);
-
-  assertLUB(nullExactDef, nullExactDef, nullExactDef, nullExactDef);
-  assertLUB(nullExactDef, none, nullDef, exactNone);
-  assertLUB(nullExactDef, nullNone, nullDef, nullExactNone);
-  assertLUB(nullExactDef, exactNone, nullExactDef, exactNone);
-  assertLUB(nullExactDef, nullExactNone, nullExactDef, nullExactNone);
-  assertLUB(nullExactDef, func, Type(Type::none), unreachable);
-  assertLUB(nullExactDef, nullFunc, Type(Type::none), unreachable);
-  assertLUB(nullExactDef, exactFunc, Type(Type::none), unreachable);
-  assertLUB(nullExactDef, nullExactFunc, Type(Type::none), unreachable);
-  assertLUB(nullExactDef, i32, Type(Type::none), unreachable);
-  assertLUB(nullExactDef, unreachable, nullExactDef, unreachable);
+  assertLUB(sup, sup, sup, sup);
+  assertLUB(sup, nullSup, nullSup, sup);
+  assertLUB(sup, exactSup, sup, exactSup);
+  assertLUB(sup, nullExactSup, nullSup, exactSup);
+  assertLUB(sup, sub, sup, sub);
+  assertLUB(sup, nullSub, nullSup, sub);
+  assertLUB(sup, exactSub, sup, exactSub);
+  assertLUB(sup, nullExactSub, nullSup, exactSub);
+  assertLUB(sup, none, sup, none);
+  assertLUB(sup, nullNone, nullSup, none);
+  assertLUB(sup, func, Type(Type::none), unreachable);
+  assertLUB(sup, nullFunc, Type(Type::none), unreachable);
+  assertLUB(sup, i32, Type(Type::none), unreachable);
+  assertLUB(sup, unreachable, sup, unreachable);
+
+  assertLUB(nullSup, nullSup, nullSup, nullSup);
+  assertLUB(nullSup, exactSup, nullSup, exactSup);
+  assertLUB(nullSup, nullExactSup, nullSup, nullExactSup);
+  assertLUB(nullSup, sub, nullSup, sub);
+  assertLUB(nullSup, nullSub, nullSup, nullSub);
+  assertLUB(nullSup, exactSub, nullSup, exactSub);
+  assertLUB(nullSup, nullExactSub, nullSup, nullExactSub);
+  assertLUB(nullSup, none, nullSup, none);
+  assertLUB(nullSup, nullNone, nullSup, nullNone);
+  assertLUB(nullSup, func, Type(Type::none), unreachable);
+  assertLUB(nullSup, nullFunc, Type(Type::none), unreachable);
+  assertLUB(nullSup, i32, Type(Type::none), unreachable);
+  assertLUB(nullSup, unreachable, nullSup, unreachable);
+
+  assertLUB(exactSup, exactSup, exactSup, exactSup);
+  assertLUB(exactSup, nullExactSup, nullExactSup, exactSup);
+  assertLUB(exactSup, sub, sup, none);
```

Comment:
About the glb being `none`, how can I see that in the lattice diagram? I'm confused as to where exact types fit in a lattice like this:
```
 any
  |
super
  |
 sub
  |
none
```

Reply:
```
  any
  |  \
super .
|   \
sub  exact super
|  \           \
.   exact sub   .
    |
    .
....
\||/
none
```

Reply:
Thanks!